pyiron_contrib/tinybase/container.py

```diff
@@ -0,0 +1,186 @@
+"""Generic Input Base Clases"""
+
+import abc
+import sys
+
+from pyiron_contrib.tinybase.storage import HasHDFAdapaterMixin
+
+from pyiron_base.interfaces.object import HasStorage
+from pyiron_atomistics.atomistics.structure.has_structure import HasStructure
+
+from ase import Atoms
+import numpy as np
+import matplotlib.pyplot as plt
+
+class StorageAttribute:
+    """
+    Create an attribute that is synced to a storage attribute.
+
+    It must be created on a class that derives from :class:`.HasStorage`.  It essentially behaves like a property that
+    writes and reads values to the underlying :attr:`.HasStorage.storage`. DataContainer of the class.  When accessing
+    this property before setting it, `None` is returned.
+
+    It's possible to modify the default value and the accepted type of values by using the builder-style :meth:`.type`
+    and :meth:`.default` methods.
+
+    >>> class MyType(HasStorage):
+    ...     a = StorageAttribute()
+    ...     b = StorageAttribute().type(int)
+    ...     c = StorageAttribute().default(list)
+    ...     d = StorageAttribute().default(lambda: 42).type(int)
```

Comment:
I find this this initialisation difficult to follow. I would prefer `StorageAttribute().default(42).type(int)` and inside `def default(self)` I'd check if it's a type or a value.

Reply:
Actually I might totally scrap the callable option and force the user to set a default value, i.e. if it's a list it must be `StorageAttribute().default([])`

Reply:
Ja, I agree it's much nicer to have to provide just the value instead of the function.  Naively it won't work to write `StorageAttribute().default([])` because every instance of this class would share the *same list instance*, i.e. if one class instance modifies its property it would be reflected in another (which we don't want).  I'll see if I can sneak in a copy somewhere though.

Reply:
Can't we just do

```python
self.default_constructor = lambda: default_constructor
```

And

```python
ret = obj.storage[self.name] = self.default_constructor()()
```

Not sure if this gives a different result though.

Reply:
Sorry, I dropped the ball here for a bit.  I'm not sure what the double call is supposed to do.  I think I'll just `copy.deepcopy()` whatever the user gives in.

Reply:
I've added both options now, and `default` does the intuitive thing now.